### ISSUE TITLE
Implement multiple algorithms to choose from besides Karp Rabin

### ISSUE THREAD

**Issue by olleharstedt** [kucherenko/jscpd#462]
**Is your feature request related to a problem? Please describe.**
The Karp Rabin - Greedy String Tiling algorithm might not necessarily be the most optimal or up-to-date algorithm to detect clones with. It's actually from 1987 (Wikipedia source), and a lot of research has happened since.

**Describe the solution you'd like**
Discussion about implementing other algorithms, and how that could be done. Which would be chosen as a candidate, and why? Risk, benefit, effort.

**Describe alternatives you've considered**
Seems like basically all available open-source clone detection tools use the Karp Rabin algo. Maybe we can improve this situation?

**Comment by kucherenko** [comment]
Hi Olle

Thank you for the request. I have some information to share about the algorithms: I've made few PoC with other algorithms and combination of algorithms (for example I tried Blums filters), as result I can say that we have most optimal algorithm in released version of `jscpd`.  If you have the algorithm what you are interested in I can try to create PoC based on it, just let me know.

Also I have one question, what you are going to achieve with new one algorithm?

**Comment by olleharstedt** [comment]
Hi!

Yeah? Did you upload any of those, @kucherenko ?

What I want to achieve:

1) No false positives on comments
2) No false positives on function calls spread out on multiple lines
3) No false positives on classes with similar properties

**Comment by olleharstedt** [comment]
Different scenarios and how they are handled by different algorithms:

![image](https://user-images.githubusercontent.com/607341/120825984-3f05b800-c55a-11eb-9da9-659a86a73611.png)

![image](https://user-images.githubusercontent.com/607341/120825960-344b2300-c55a-11eb-8db6-13f632b799b6.png)

https://reader.elsevier.com/reader/sd/pii/S0167642309000367?token=4CEC8ED633248046C8ECF054461C1077C4931D0142A7EA14087C2C27E8764BA65CB89CE01575246C4DDE06E64302B9D1&originRegion=eu-west-1&originCreation=20210604152708

**Comment by kucherenko** [comment]
Thank you for the docs, in jscpd you can try to customize detection process with modes, in modes you can filter or modify all tokens before starting detection

**Comment by olleharstedt** [comment]
If you wanna support multiple languages, it's easiest with text or token based approaches, I guess, and not graph based. I'm looking a bit at the algorithm here, which is available as open-source (2015): https://www.cqse.eu/fileadmin/content/news/publications/2009-do-code-clones-matter.pdf

Guess I'll close the issue now, and open a PR if I ever see some progress. :) Thanks for the feedback!

**Comment by olleharstedt** [comment]
I *think* I made some progress.

Have a look at the clone in this file: 


https://github.com/LimeSurvey/LimeSurvey/blob/master/application/models/QuestionTheme.php#L363

https://github.com/LimeSurvey/LimeSurvey/blob/master/application/models/QuestionTheme.php#L836

```php
if (\PHP_VERSION_ID < 80000) {
    $bOldEntityLoaderState = libxml_disable_entity_loader(true);
}
$sQuestionConfigFilePath = App()->getConfig('rootdir') . DIRECTORY_SEPARATOR . $pathToXML . DIRECTORY_SEPARATOR . 'config.xml';
if (!file_exists($sQuestionConfigFilePath)) {
    throw new Exception(gT('Extension configuration file is not valid or missing.'));
}
$sQuestionConfigFile = file_get_contents($sQuestionConfigFilePath);  // @see: Now that entity loader is disabled, we can't use simplexml_load_file; so we must read the file with file_get_contents and convert it as a string
```

and

```php
if (\PHP_VERSION_ID < 80000) {
    $bOldEntityLoaderState = libxml_disable_entity_loader(true);
}
$sQuestionConfigFilePath = App()->getConfig('rootdir') . DIRECTORY_SEPARATOR . $sConfigPath;
if (!file_exists($sQuestionConfigFilePath)) {
    throw new Exception('Found no config.xml file at ' . $sQuestionConfigFilePath);
}
$sQuestionConfigFile = file_get_contents($sQuestionConfigFilePath);  // @see: Now that entity loader is disabled, we can't use simplexml_load_file; so we must read the file with file_get_contents and convert it as a string
```

As you can see it's slightly edited. The new algorithm I'm copying has a parameter called **edit distance**, which makes it possible to detect clones with slight differences like these, even on token level. I did not manage to detect the whole 8-lines clone with jscpd nor phpcpd. Have a go if you like.

My repo is here: https://github.com/olleharstedt/suffixtree

Main entry point:

```java
/**
 * Finds all clones in the string (List) used in the constructor.
 * @param minLength the minimal length of a clone
 * @param maxErrors the maximal number of errors/gaps allowed
 * @param headEquality the number of elements which have to be the same at the beginning of a clone
 */
public void findClones(int minLength, int maxErrors, int headEquality) throws ConQATException {
```

Ping me if you want instructions on how to run it. It's not tidy at all, I just bashed on it until it worked, but the core of the algorithm is not changed.

**Comment by kucherenko** [comment]
Thank you,  it is looks interesting, as disadvantage of the solution I can say that I should implement the algorithm for each of supported languages. Also I can suggest you to investigate jsinspect project - https://github.com/danielstjules/jsinspect, jsinspect identify code with a similar structure.

**Comment by olleharstedt** [comment]
> as disadvantage of the solution I can say that I should implement the algorithm for each of supported languages

No, the solution is still token-based! That's the good thing. :) But instead of comparing a hash, they build suffix-trees and compare. I don't know the details so much, but it was no problem to feed it PHP tokens instead of whatever they used. Can I assume you have a token class you convert all languages to?

Will check the link, thanks!

**Comment by kucherenko** [comment]
thank you, will try

**Comment by olleharstedt** [comment]
This topic might be moot due to license issues. The license for the new algorithm is Apache 2.0, but you're using MIT. IIRC, you'd have to re-license to Apache 2.0 too if you'd want to include it. :( Or GPL v3

https://en.wikipedia.org/wiki/Apache_License

**Comment by kucherenko** [comment]
Closing with what landed instead of a second exact-match algorithm. The Rabin-Karp core stays, and three opt-in mechanisms sit on top of it, all implemented from scratch under the project's MIT license (so the licensing concern above does not apply):

- **Type-2 normalization** (#998, #1019): `--ignore-identifiers`, `--ignore-literals`, `--ignore-annotations` hash token classes to placeholders so renamed or re-literaled copies match.
- **Gap-tolerant merging** (#999 stage 1, #1020): `--max-gap-lines N` merges the fragments around an inserted or removed line into one near-miss clone, which is the fragmentation this issue describes.
- **Function-level similarity for JS/TS** (#999 stage 2, #1021): `--similarity RATIO` compares whole functions by AST node-type shingles with MinHash + LSH candidates and a weighted Jaccard score, i.e. a token-independent algorithm alongside the windowed one.

Every mechanism is off by default, so existing runs are unchanged. Other languages plug into the similarity engine through a `FunctionExtractor` per grammar. Anything beyond that (tree edit distance, suffix trees) can be a fresh issue against the new plug-in point.
